wxQt: wxEVT_TEXT_ENTER processing revisited

AliKet opened the pull request:
@vadz sorry for the noise but I think this one is better than the previous
(now merged!) one.

I really want `wxSpinCtrl` to take advantage of this, but unfortunately it doesn't derive from `wxTextEntry`.

Commit:
Partially revert commits 119f24761c and a7cec904c8

Instead of commits a7cec904c8 (Handle wxEVT_TEXT_ENTER properly in wxQt
added in 2022-11-03) and 119f24761c, we'll handle wxEVT_TEXT_ENTER more
properly (in the upcomming commits) without the need to toggle the default
button at all.

Commit:
Add a hack to wxQtEventSignalHandler<> for wxEVT_TEXT_ENTER handling

wxQtEventSignalHandler<> derived classes (wxTextCtrl-like controls)
only have to override wxQtSignalHandler<>::GetValueForProcessEnter()
to get wxEVT_TEXT_ENTER event for free.

AliKet commented:
Why double `WXUNUSED(WXUNUSED(...))` here ?

https://github.com/wxWidgets/wxWidgets/blob/222cf952f4618d4b907193359753e841a854702b/src/qt/spinctrl.cpp#L26-L31

vadz commented:
Sorry, I'm a bit confused here: so we _can_, finally, prevent Qt from closing the dialog on pressing Enter? I.e. what has changed to allow removing the hack with `g_defaultButton`?

Also, do we really want to deal with SFINAE here when we could simply have `virtual bool QtProcessEnter() { return false; }` in the base class and just override it in the classes that need to do something else? AFAICS this would be much more straightforward and we really don't care about the cost of an extra virtual function call when dispatching key events (people don't type fast enough for this to be relevant).

Finally, the double `WXUNUSED()` is just a mistake IMO.

AliKet commented:
> Sorry, I'm a bit confused here: so we _can_, finally, prevent Qt from closing the dialog on pressing Enter?

By good luck yes!

> I.e. what has changed to allow removing the hack with `g_defaultButton`?
> 

Previously, we request handling `wxEVT_TEXT_ENTER` from inside `wxQtComboBox::activated()` and `wxQtLineEdit::returnPressed()` signals, and apparently (or in fact, because this is what actually happen) the event (generated by pressing enter) is propagated to the dialog which in turn looks for the default button to forward the event to, and there is no way to prevent this from happening ,if `wxEVT_TEXT_ENTER` has been processed and no `Skip()` was called, other than temporarily toggle the default button off.

Fortunately the above scenario won't happen if the control really handle `wxEVT_TEXT_ENTER` in `keyPressEvent()` and the event was `accept()`ed

> Also, do we really want to deal with SFINAE here when we could simply have `virtual bool QtProcessEnter() { return false; }` in the base class and just override it in the classes that need to do something else? AFAICS this would be much more straightforward and we really don't care about the cost of an extra virtual function call when dispatching key events (people don't type fast enough for this to be relevant).
> 

I chose this approach **1)** for performance reason as you've already guessed and **2)** to avoid code duplication as much as possible.
If these two arguments are not convincing then let's add `virtual bool QtProcessEnter() { return false; }` and the derived classes override this at their will.

> Finally, the double `WXUNUSED()` is just a mistake IMO.

vadz commented:
Thanks for the explanation! The new approach definitely seems much better (now that I understand why it works).

As for the choice of approach: I really don't think we care that much about performance of handling "Enter". Using a virtual function seems like a much simpler solution and we could avoid code duplication simply by having a helper (inline) function that could be called from the overridden version. So yes, I think it'd be preferable to do it like this to keep things simple and easy to modify.

TIA!

AliKet commented:
@vadz I hope you like the new implementation!
FYI it works as expected with wxTextCtrl, wxComboBox and wxSpinCtrl[Double]

Commit:
Use wxSpinCtrlBase instead of wxControl in wxQtSpinBoxBase definition

This fixes compilation error introduced by parent commit d613d4e680
(Add a hack to wxQtEventSignalHandler<> for wxEVT_TEXT_ENTER handling)
because wxQtSpinBoxBase<Widget>::GetValueForProcessEnter() override calls
GetTextValue() which is a member of wxSpinCtrlBase and not wxControl.

Other than that, no change in behavior because of it.

vadz commented:
Looks good to me, thanks!

AliKet commented:
Please don't merge, I have some pending commits that I want to push soon,
TIA!

Commit:
Clean up leftover connection to the already removed signal

Commit:
Stop handling wxEVT_TEXT_ENTER twice

Can happen if ProcessTextEnter() returns false due to wxEVT_TEXT_ENTER handler
calling Skip().

So change ProcessTextEnter() to HandleKeyPressEvent() and emit wxEVT_TEXT_ENTER
from there if necessary.

Commit:
Make wxQtSignalHandler a non-template class

To handle wx events the template parameter Handler should derive from
wxEvtHandler. This requirement is enforced now by making wxQtSignalHandler
a non-template class with an explicit wxEvtHandler member variable.

As a side effect, this will tremendously eliminate a bunch of unnecessary
template instantiation and reduce code bloat as a result.

Commit:
Fixup for classes deriving from wxQtSignalHandler

Also clear the confusion about the handler:
- wxToolBar or wxToolBarTool for wxQtToolButton
- wxMenu or wxMenuItem for wxQtAction
Because wxToolBar and wxMenu which are for event handling not
wxToolBarTool or wxMenuItem which are wxObjects only.

AliKet commented:
The samples (including the widgets sample) work correctly with these changes
but a double check or even better, testing this with your application (if you are a `wxQt` guy of course) is really appreciate, TIA!
